agent/handlers/agentapi/taskprotection/v1/handlers/handlers.go

```diff
@@ -257,22 +222,108 @@ func GetTaskProtectionHandler(state dockerstate.TaskEngineState, credentialsMana
+	taskRoleCredential, ok := credentialsManager.GetTaskCredentials(task.GetCredentialsID())
+	if !ok {
+		return nil, http.StatusForbidden, ecs.ErrCodeAccessDeniedException, errors.New("Invalid Request: no task IAM role credentials available for task")
+	}
+	taskCredential := taskRoleCredential.GetIAMRoleCredentials()
+	cfg := aws.NewConfig().
+		WithCredentials(awscreds.NewStaticCredentials(taskCredential.AccessKeyID,
+			taskCredential.SecretAccessKey,
+			taskCredential.SessionToken)).
+		WithRegion(factory.Region).
+		WithHTTPClient(httpclient.New(ecsclient.RoundtripTimeout, factory.AcceptInsecureCert)).
+		WithEndpoint(factory.Endpoint)
+
+	ecsClient := ecs.New(session.Must(session.NewSession()), cfg)
+	return ecsClient, http.StatusOK, "", nil
+}
+
+// Helper function to parse error to get ErrorCode, ExceptionMessage, HttpStatusCode, RequestID.
+// RequestID will be empty if the request is not able to reach AWS
+func getErrorCodeAndStatusCode(err error) (string, string, int, *string) {
+	msg := err.Error()
+	// The error is a Generic AWS Error with Code, Message, and original error (if any)
+	if awsErr, ok := err.(awserr.Error); ok {
+		// The error is an AWS service error occurred
+		msg = awsErr.Message()
+		if reqErr, ok := err.(awserr.RequestFailure); ok {
+			reqId := reqErr.RequestID()
+			return awsErr.Code(), msg, reqErr.StatusCode(), &reqId
+		} else {
+			logger.Error(fmt.Sprintf("got an exception that does not implement RequestFailure interface but is an aws error. This should not happen, return statusCode 500 for whatever errorCode. Original err: %v.", err))
+			return awsErr.Code(), msg, http.StatusInternalServerError, nil
+		}
+	} else {
+		logger.Error(fmt.Sprintf("non aws error received: %v", err))
+		return ecs.ErrCodeServerException, msg, http.StatusInternalServerError, nil
+	}
+}
+
+// Helper function for finding task for the request
+func getTaskFromRequest(state dockerstate.TaskEngineState, r *http.Request) (*apitask.Task, int, string, error) {
+	taskARN, err := v3.GetTaskARNByRequest(r, state)
+	if err != nil {
+		logger.Error("Failed to find task ARN for task protection request", logger.Fields{
+			loggerfield.Error: err,
+		})
+		return nil, http.StatusNotFound, ecs.ErrCodeResourceNotFoundException, errors.New("Invalid request: no task was found")
+	}
+
+	task, found := state.TaskByArn(taskARN)
+	if !found {
+		logger.Critical("No task was found for taskARN for task protection request", logger.Fields{
+			loggerfield.TaskARN: taskARN,
+		})
+		return nil, http.StatusInternalServerError, ecs.ErrCodeServerException, errors.New("Failed to find a task for the request")
+	}
+
+	return task, http.StatusOK, "", nil
+}
+
+// generateErrorResponse generates an error in *type.ErrorResponse format for Agent input validations failures and exceptions
+func generateErrorResponse(arn string, code string, message string) *types.ErrorResponse {
```

Comment:
Should we move this constructor to `types` package?

Reply:
sure

Reply:
I initially put it here since it is not a standard constructor but just a helper function to avoid initialization in function call, but I guess I can change it to a constructor.